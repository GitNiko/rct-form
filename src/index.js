import React from "react";
import { reduxForm, Field } from "redux-form";

const fields = {};
const setField = (type, component) => (fields[type] = component);
/**
 * 
 * @param {string} type 
 * @return {React.Component}
 */
const getField = type => fields[type];

export const rctField = type => InnerComponent => {
  class InnerWrapper extends React.Component {
    render() {
      return <Field {...this.props} component={InnerComponent} />;
    }
  }
  setField(type, InnerWrapper);
  return InnerWrapper;
};

export const rctForm = ReactComponent =>
  class Wrapper extends React.Component {
    constructor(props) {
      super(props);
      const { formProp, validate } = props;
      const FormComponent = reduxForm({ form: formProp.formName, validate })(
        ReactComponent
      );
      this.temp = p => <FormComponent {...p} />;
    }

    render() {
      const { formProp } = this.props;
      if (!formProp || !formProp.formName) {
        console.warn("formName is invalid, formProp:", formProp);
        return null;
      }
      const newProps = { getField, ...this.props };
      return this.temp(newProps);
    }
  };
