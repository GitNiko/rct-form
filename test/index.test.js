import React from "react";
import { createStore, combineReducers } from "redux";
import { Provider, connect } from "react-redux";
import { reducer as formReducer, formValueSelector } from "redux-form";
import { rctForm, rctField } from "../src";
import renderer from "react-test-renderer";

const formDSL = {
  formName: "creditAdd",
  sections: [
    {
      fields: [
        {
          fieldType: "selectItem",
          name: "user.customer",
          label: "选择客户",
          placeholder: "请选择客户"
        },
        {
          fieldType: "inputItem",
          name: "user.username",
          label: "账号",
          placeholder: "请输入账号",
          type: "text"
        },
        {
          fieldType: "inputItem",
          name: "user.password",
          label: "密码",
          placeholder: "请输入密码",
          type: "password"
        },
        {
          fieldType: "inputItem",
          name: "user.phone",
          label: "电话",
          placeholder: "请输入电话号码",
          type: "phone",
          dependencies: [
            {
              type: "and",
              rules: [
                {
                  fieldName: "user.username",
                  type: "regular",
                  value: "admin"
                }
              ]
            }
          ]
        }
      ]
    }
  ]
};

class InputItemField extends React.Component {
  onChange = e => {
    const { input } = this.props;
    input.onChange(e.target.value);
  };

  render() {
    const {
      type,
      placeholder,
      editable,
      disabled,
      clear,
      maxLength,
      extra,
      label,
      labelNumber,
      updatePlaceholder,
      moneyKeyboardAlign,
      meta: { error, touched },
      input
    } = this.props;
    return (
      <div>
        {label}:
        <input
          type={type}
          placeholder={placeholder}
          value={input.value ? input.value : ""}
          disabled={disabled}
          maxLength={maxLength}
          onChange={this.onChange}
        />
      </div>
    );
  }
}

class MyForm extends React.Component {
  componentDidMount() {
    this.props.bindSubmit(this.props.submit);
  }
  isNeed(formName, field) {
    const dependencies = field.dependencies;
    let need = true;
    if (
      dependencies &&
      Array.isArray(dependencies) &&
      dependencies.length > 0
    ) {
      //遍历所有dependence,最外层为【且】，只要有一个不满足，则不显示
      for (const dependence of dependencies) {
        if (!this.validateDependence(dependence, formName)) {
          need = false;
          break;
        }
      }
    }
    return need;
  }

  validateDependence = ({ type, rules }, formName) => {
    let result = false;
    if (rules && Array.isArray(rules) && rules.length > 0) {
      for (const rule of rules) {
        switch (rule.type) {
          case "regular":
            const selector = formValueSelector(formName);
            const formValue = selector(this.props.state, rule.fieldName);
            const regex = new RegExp(rule.value);
            result = regex.test(String(formValue));
            break;
          default:
            break;
        }
        if (type === "or" && result) {
          break;
        }
      }
    }
    return result;
  };
  render() {
    const { getField, formProp } = this.props;
    // DSL resolve
    const cmps = formProp.sections.map((section, index) => (
      <div key={index}>
        {section.fields.map(field => {
          const CMP = getField(field.fieldType);
          if (CMP) {
            if (this.isNeed(formProp.formName, field)) {
              return <CMP {...field} key={field.name} />;
            } else {
              return null;
            }
          } else {
            return null;
          }
        })}
      </div>
    ));
    return <div>{cmps}</div>;
  }
}

describe("form init", () => {
  const Form = rctForm(MyForm);
  const field = rctField("inputItem")(InputItemField);
  const store = createStore(combineReducers({ form: formReducer }));
  const bindSubmit = submitFn => null;

  it("render form", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Form formProp={formDSL} bindSubmit={bindSubmit} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("render form of empty name", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Form formProp={{}} bindSubmit={bindSubmit} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
