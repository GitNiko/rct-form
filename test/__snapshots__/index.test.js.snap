// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form init render form 1`] = `
<div>
  <div>
    <div>
      账号
      :
      <input
        disabled={undefined}
        maxLength={undefined}
        onChange={[Function]}
        placeholder="请输入账号"
        type="text"
        value=""
      />
    </div>
    <div>
      密码
      :
      <input
        disabled={undefined}
        maxLength={undefined}
        onChange={[Function]}
        placeholder="请输入密码"
        type="password"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`form init render form of empty name 1`] = `null`;
